### PR TITLE
feat(platform): Add Host-scoped credentials support for blocks HTTP requests

### PR DESCRIPTION
**Pull request by majdyz**
Currently, we don't have a secure way to pass Authorization headers when calling the `SendWebRequestBlock`.
This hinders the integration of third-party applications that do not yet have native block support.

### Changes 🏗️

Add Host-scoped credentials support for the newly introduced SendAuthenticatedWebRequestBlock.

image
image

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  
  - [x] Uses `https://api.openai.com/v1/images/edits` through SendWebRequestBlock by passing the api-key through host-scoped credentials.

### CODE REVIEW

File: autogpt_platform/backend/backend/blocks/http.py [hunk 1]
```diff
@@ -169,3 +214,61 @@ async def run(
...
+            self,
+            id="7f4f8b2a-c3d1-4e5f-8a9b-2c3d4e5f6a7b",
+            description="Make an authenticated HTTP request with host-scoped credentials (JSON / form / multipart).",
+            categories={BlockCategory.OUTPUT},
+            input_schema=SendAuthenticatedWebRequestBlock.Input,
+            output_schema=SendWebRequestBlock.Output,
+            test_credentials=TEST_CREDENTIALS,
+        )
+
+    async def run(  # type: ignore[override]
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/http.py:224]: ??

File: autogpt_platform/backend/backend/executor/manager.py [hunk 2]
```diff
@@ -209,11 +209,19 @@ async def execute_node(
...
+
+        # Only acquire credentials if an ID is provided and not empty
+        if credentials_meta and credentials_meta.id and credentials_meta.id.strip():
+            credentials, creds_lock = await creds_manager.acquire(
+                user_id, credentials_meta.id
+            )
+            extra_exec_kwargs[field_name] = credentials
+        else:
+            # No credentials provided - pass None to the block
+            extra_exec_kwargs[field_name] = None
```
Comment (ntindle) [inline autogpt_platform/backend/backend/executor/manager.py]: Is the assumption that this cannot fail to inject here depended on by anything else?

**Review by ntindle — APPROVED** [review]
Backend lgtm

File: autogpt_platform/backend/backend/blocks/http.py (hunk 1 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/blocks/http.py:224]: The parent doesn't have credentials as part of the param

File: autogpt_platform/backend/backend/executor/manager.py (hunk 2 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/executor/manager.py]: I was going to provide optional credential initially, now it doesn't matter let me revert this

**Comment by AutoGPT-Agent** [comment]
Thank you for implementing host-scoped credentials support for blocks HTTP requests! This is a valuable feature that will enhance the platform's integration capabilities.

Your PR description is clear, and the test plan is specific which is great. The screenshots provide helpful visual context for the feature implementation.

Since this PR modifies files in the backend/data directory (graph.py and model.py), please ensure that:

1. Any new or modified functions in these files correctly pass and check the user_id parameter to maintain proper access controls
2. If there are any exceptions to this pattern, please add an explanation in the PR description

Also, given the security implications of handling credentials, consider adding some notes about:

- How credentials are securely stored
- Any access control mechanisms implemented to prevent unauthorized access to stored credentials
- How credentials are passed securely to the HTTP requests

This would help reviewers better understand the security considerations that went into this implementation.

Once these points are addressed or confirmed, this PR looks ready to merge.

**Review by ntindle — DISMISSED** [review]
Approved for backend stuf

**Comment by AutoGPT-Agent** [comment]
Thank you for this PR implementing Host-scoped credentials for blocks HTTP requests! This will definitely improve the integration capabilities with third-party applications.

A few notes:

1. The PR title should use our defined scope format. Instead of `feat(backend;frontend)`, it should be something like `feat(platform/backend,platform/frontend)` to align with our conventions.

2. Since there are changes to files in `backend/data/`, could you confirm that any added/modified functions properly handle user_id checks? This is important for security to ensure proper access control.

3. I see this PR has a conflicts label - please resolve these conflicts before it can be merged.

Otherwise, the PR looks well-documented with a clear purpose, implementation details, and test plan. The feature will be a valuable addition for secure third-party integrations.

File: autogpt_platform/frontend/package.json [hunk 3]
```diff
@@ -133,5 +133,5 @@
       "public"
     ]
   },
-  "packageManager": "pnpm@10.11.1+sha256.211e9990148495c9fc30b7e58396f7eeda83d9243eb75407ea4f8650fb161f7c"
+  "packageManager": "pnpm@10.12.3"
```
Comment (Pwuts) [inline autogpt_platform/frontend/package.json]: this is a security downgrade right? (from hashed to non-hashed version reference)

File: autogpt_platform/backend/test/data/test_models.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
+import pytest
+from pydantic import SecretStr
+
+from backend.data.model import HostScopedCredentials
+
+
+class TestHostScopedCredentials:
```
Comment (Pwuts) [inline autogpt_platform/backend/test/data/test_models.py]: pls put the tests in the module level like in all other test files

File: autogpt_platform/backend/backend/blocks/http.py (hunk 1 above)
Comment (Pwuts) [reply autogpt_platform/backend/backend/blocks/http.py:224]: That's allowed without `# type: ignore` because of the `**kwarg`

File: autogpt_platform/backend/backend/blocks/http.py [hunk 5]
```diff
@@ -169,3 +207,61 @@ async def run(
...
+class SendAuthenticatedWebRequestBlock(SendWebRequestBlock):
+    class Input(SendWebRequestBlock.Input):
+        credentials: HttpCredentials = SchemaField(
+            description="HTTP host-scoped credentials for automatic header injection"
+        )
+
+    def __init__(self):
+        Block.__init__(
+            self,
+            id="7f4f8b2a-c3d1-4e5f-8a9b-2c3d4e5f6a7b",
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/blocks/http.py]: `-8a9b-2c3d4e5f6a7b`

AI-generated UUID detected. Seems like a bad practice to me?

File: autogpt_platform/backend/backend/data/model.py [hunk 6]
```diff
@@ -240,13 +241,61 @@ def auth_header(self) -> str:
         return f"Basic {base64.b64encode(f'{self.username.get_secret_value()}:{self.password.get_secret_value()}'.encode()).decode()}"
 
 
+class HostScopedCredentials(_BaseCredentials):
+    type: Literal["host_scoped"] = "host_scoped"
+    host: str = Field(description="The host/URI pattern to match against request URLs")
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model.py:246]: I'd add a field validator for this to ensure it's a valid URL or a pattern in the format that we need

File: autogpt_platform/backend/backend/data/model.py [hunk 7]
```diff
@@ -240,13 +241,61 @@ def auth_header(self) -> str:
...
+    )
+
+    @field_serializer("headers")
+    def serialize_headers(self, headers: dict[str, SecretStr]) -> dict[str, str]:
+        """Serialize headers by extracting secret values."""
+        return {key: value.get_secret_value() for key, value in headers.items()}
+
+    def get_headers_dict(self) -> dict[str, str]:
+        """Get headers with secret values extracted."""
+        return {key: value.get_secret_value() for key, value in self.headers.items()}
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model.py]: Can these be deduplicated?

File: autogpt_platform/backend/backend/data/graph.py [hunk 8]
```diff
@@ -282,36 +283,49 @@ def aggregate_credentials_inputs(
         Returns:
             dict[aggregated_field_key, tuple(
                 CredentialsFieldInfo: A spec for one aggregated credentials field
+                    (now includes sibling_inputs from matching nodes)
                 set[(node_id, field_name)]: Node credentials fields that are
                     compatible with this aggregated field spec
             )]
         """
+        # First collect all credential field data with input defaults
+        credential_data = []
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/graph.py]:
```suggestion
        nodes_credential_fields = []
```

File: autogpt_platform/frontend/src/components/integrations/credentials-input.tsx [hunk 9]
```diff
@@ -148,14 +151,69 @@ export const CredentialsInput: FC<{
...
+      return urlObj.hostname;
+    } catch {
+      return null;
+    }
+  };
 
-    if (credentials.savedCredentials.length === 1)
-      return credentials.savedCredentials[0];
+  // Get current host from siblingInputs
+  const currentUrl = siblingInputs?.url;
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/components/integrations/credentials-input.tsx]: The `CredentialsField` has a `discriminator` parameter for this purpose; this relation with the `url` input shouldn't be hard-coded here. You may have to adjust the logic in `useCredentials` a little bit, but that's where this discrimination and pre-filtering of credentials currently happens.

File: autogpt_platform/frontend/package.json (hunk 3 above)
Comment (0ubbe) [reply autogpt_platform/frontend/package.json]: Corepack hashes the versions for security:
- https://github.com/nodejs/corepack?tab=readme-ov-file#when-authoring-packages

Given `pnpm@10.12.3` was released yesterday, is likely corepack's built-in manifest. Usually it takes around 2-3 days, so I'd say let's wait for now 👍🏽

File: autogpt_platform/frontend/src/app/(platform)/profile/(user)/integrations/page.tsx [hunk 10]
```diff
@@ -5,6 +5,7 @@ import { useCallback, useContext, useEffect, useMemo, useState } from "react";
 import { useToast } from "@/components/ui/use-toast";
 import { IconKey, IconUser } from "@/components/ui/icons";
 import { Trash2Icon } from "lucide-react";
+import { FaKey } from "react-icons/fa";
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/app/(platform)/profile/(user)/integrations/page.tsx]: I have agreed with Olivia to settle on **Phospor icons**, so we don't have like 5 icons libraries on the app 😂 

This one should be good: https://phosphoricons.com/?q=key

```tsx
import { KeyIcon } from ""@phosphor-icons/react/dist/ssr"
```

File: autogpt_platform/frontend/src/components/integrations/credentials-provider.tsx [hunk 11]
```diff
@@ -173,6 +183,22 @@ export default function CredentialsProvider({
...
+    ): Promise<CredentialsMetaResponse> => {
+      const credsMeta = await api.createHostScopedCredentials({
+        provider,
+        ...credentials,
+      });
+      addCredentials(provider, credsMeta);
+      return credsMeta;
+    },
+    [api, addCredentials],
+  );
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/integrations/credentials-provider.tsx]: What happens if this API call fails? Are we handling the error on the `BackendAPI` client? We should at least show a toast 💭

File: autogpt_platform/frontend/src/components/integrations/credentials-input.tsx [hunk 12]
```diff
@@ -695,6 +796,223 @@ export const UserPasswordCredentialsModal: FC<{
   );
 };
 
+export const HostScopedCredentialsModal: FC<{
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/integrations/credentials-input.tsx]: It is a good practice to stick to 1 component per file. Can we move this to a separate file in the same directory? 📁

File: autogpt_platform/frontend/src/components/integrations/credentials-input.tsx [hunk 13]
```diff
@@ -148,14 +151,69 @@ export const CredentialsInput: FC<{
...
-    if (!credentials || !("savedCredentials" in credentials)) return null;
+  // Helper function to extract host from URL
+  const getHostFromUrl = (url: string): string | null => {
+    try {
+      const urlObj = new URL(url);
+      return urlObj.hostname;
+    } catch {
+      return null;
+    }
+  };
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/integrations/credentials-input.tsx]: These type of helpers can go into `src/lib/utils/url.ts` for example, as it is likely we will need them in the future in other contexts 👍🏽

**Review by 0ubbe — COMMENTED** [review]
Looking good 👍🏽 , left some suggestions 💬

File: autogpt_platform/backend/backend/blocks/http.py (hunk 1 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/blocks/http.py:224]: Also the input type is different
```
Method "run" overrides class "SendWebRequestBlock" in an incompatible manner
  Parameter 2 type mismatch: base parameter is type "Input", override parameter is type "Input"
    "backend.blocks.http.SendWebRequestBlock.Input" is incompatible with "backend.blocks.http.SendAuthenticatedWebRequestBlock.Input"
```
idk why it's incompatible even though the later one extending the former

File: autogpt_platform/frontend/src/components/integrations/credentials-input.tsx (hunk 13 above)
Comment (majdyz) [reply autogpt_platform/frontend/src/components/integrations/credentials-input.tsx]: Updated

File: autogpt_platform/frontend/src/components/integrations/credentials-input.tsx (hunk 12 above)
Comment (majdyz) [reply autogpt_platform/frontend/src/components/integrations/credentials-input.tsx]: You are right, I see there are multiple other modal components here, so I decoupled each of them into different files here.
The PR gets bigger though, I hope you don't mind.

**Comment by AutoGPT-Agent** [comment]
Thanks for this PR implementing Host-scoped credentials support for blocks HTTP requests! This is a valuable addition that will help with third-party integrations.

The PR description and checklist look good. I appreciate the screenshots showing the UI implementation and the clear test plan.

A few things to consider:

1. Since this PR is labeled with "Possible security concern" and touches credential handling, could you elaborate on the security measures implemented to ensure credentials are properly protected?

2. For the modified functions in the backend/data/*.py files, could you confirm that user_id checks are properly implemented where needed?

3. The PR introduces a new block type (SendAuthenticatedWebRequestBlock). Is this documented somewhere for users to understand how to use it correctly?

4. Could you provide some more details on how the host-scoped credentials are stored and managed securely?

Once these points are addressed, the PR should be ready for a more detailed code review. Great work on implementing this feature!

File: autogpt_platform/frontend/src/components/integrations/credentials-provider.tsx (hunk 11 above)
Comment (majdyz) [reply autogpt_platform/frontend/src/components/integrations/credentials-provider.tsx]: I have added the toast, I see all the api call here in this file don't call toast on failure, so I added each of them here.
Again, the PR gets bigger though, hehe

File: autogpt_platform/frontend/src/app/(platform)/profile/(user)/integrations/page.tsx (hunk 10 above)
Comment (majdyz) [reply autogpt_platform/frontend/src/app/(platform)/profile/(user)/integrations/page.tsx]: Updated

File: autogpt_platform/backend/backend/data/graph.py (hunk 8 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/data/graph.py]: Updated

File: autogpt_platform/backend/backend/data/model.py (hunk 7 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/data/model.py]: Updated

File: autogpt_platform/backend/backend/data/model.py (hunk 6 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/data/model.py:246]: You can't validate host, because `localhost` or any random string also a legit host.

File: autogpt_platform/backend/backend/blocks/http.py (hunk 5 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/blocks/http.py]: Yeah you are right, updated

File: autogpt_platform/backend/test/data/test_models.py (hunk 4 above)
Comment (majdyz) [reply autogpt_platform/backend/test/data/test_models.py]: Done, along with all other tests, turns out we are not running those tests on CI, so I moved everything out and fixed the broken ones.

File: autogpt_platform/frontend/package.json (hunk 3 above)
Comment (majdyz) [reply autogpt_platform/frontend/package.json]: Okay reverted

**Comment by AutoGPT-Agent** [comment]
Thank you for implementing host-scoped credentials support for blocks HTTP requests. This is a valuable addition that will help with third-party integrations. I have a few concerns that need to be addressed before this can be merged:

1. **Backend Data Files**: You've modified several files in `backend/data/` including execution.py, graph.py, integrations.py, and model.py. For security reasons, we need to ensure that any added or changed functions correctly pass and validate the user_id parameter. Could you please confirm this is the case or explain why user_id validation isn't needed for these specific changes?

2. **Security Review**: Given this PR deals with credential handling, it would be beneficial to provide more details about the security measures implemented to protect these credentials, especially how they're stored, accessed, and used in HTTP requests.

3. **Implementation Details**: While the screenshots provide a visual of the UI, it would be helpful to have a brief explanation of the architecture for how the host-scoped credentials work, including:
   - How credentials are stored securely
   - How they're associated with specific hosts
   - The authorization flow when a block needs to use these credentials

4. **Test Coverage**: The test plan mentions using the OpenAI API as a test case. Could you provide more comprehensive test scenarios covering different edge cases like:
   - Handling of invalid credentials
   - Security boundaries between different users' credentials
   - Error handling scenarios

Once these points are addressed, we can proceed with a more detailed review of the implementation.

**Comment by AutoGPT-Agent** [comment]
Thank you for this PR implementing Host-scoped credentials support for blocks HTTP requests. This is a valuable addition that will enhance security when integrating with third-party APIs.

A few points to consider:

1. I notice there are changes to several files in `backend/data/` including `execution.py`, `graph.py`, `integrations.py`, and `model.py`. Could you ensure that any new or modified functions in these files properly pass and check the user_id for authorization purposes?

2. Since this PR deals with credential handling, have you considered any potential security implications? The "Possible security concern" label suggests this might need additional review from a security perspective.

3. The test plan mentions using OpenAI's API as a test case. Have you considered adding more comprehensive tests that cover various edge cases for the credential management?

4. Could you clarify if there are any backward compatibility concerns with existing blocks or workflows?

Overall, the PR looks promising and well-structured. With a few clarifications on the points above, particularly regarding proper user_id handling in the backend/data files, this should be ready for merging.

File: autogpt_platform/frontend/src/components/integrations/credentials-input.tsx (hunk 9 above)
Comment (majdyz) [reply autogpt_platform/frontend/src/components/integrations/credentials-input.tsx]: I've update to use the discriminator, however for uri, the credentials can't simply identified by a provider, there is a uri, we the discriminator here end up being uri / host, and the dicriminator_value has to be exposed and the aggregation need to be adjusted according to this.

**Comment by AutoGPT-Agent** [comment]
Thank you for your contribution! The PR looks generally well-structured and addresses an important security concern by adding host-scoped credentials for HTTP requests.

A few things to check before we can merge:

1. The changes to files in `backend/data/` need to be reviewed for proper user_id handling:
   - Please ensure that any new or modified functions in `execution.py`, `graph.py`, `integrations.py`, and `model.py` are correctly handling user authentication and authorization.

2. While the test plan mentions testing with OpenAI's API, could you provide a bit more detail about how you verified the security aspects of this implementation? Specifically:
   - How are credentials stored securely?
   - How is access to these credentials controlled/limited?
   - What prevents credential leakage across different users/organizations?

3. Did you consider any potential security implications or edge cases with the host-scoped credential implementation?

Once these items are addressed, this PR should be ready for final review and merging. Thanks for improving the security posture of the platform!

File: autogpt_platform/backend/backend/conftest.py [hunk 14]
```diff
@@ -6,6 +6,8 @@
 
 from backend.util.logging import configure_logging
 
+os.environ["ENABLE_AUTH"] = "false"
```
Comment (ntindle) [inline autogpt_platform/backend/backend/conftest.py:9]: we shouldnt depend on this ++ should remove it

Comment (majdyz) [reply autogpt_platform/backend/backend/conftest.py:9]: We probably don't, it's a refactoring remains and keeping what's already there

File: autogpt_platform/frontend/src/lib/utils/url.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Extracts the hostname from a URL string.
+ * @param url - The URL string to extract the hostname from
+ * @returns The hostname if valid, null if invalid
+ */
+export const getHostFromUrl = (url: string): string | null => {
+  try {
+    if (!url.startsWith("http://") && !url.startsWith("https://")) {
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/lib/utils/url.ts:8]:
```suggestion
    if (!url.contains("://")) {
```

File: autogpt_platform/frontend/src/hooks/useCredentials.ts [hunk 16]
```diff
@@ -33,12 +38,16 @@ export default function useCredentials(
...
-      credsInputSchema.discriminator_mapping![
-        getValue(credsInputSchema.discriminator, nodeInputValues)
-      ]) ||
-    null;
+  const discriminatorValue = [
+    credsInputSchema.discriminator
+      ? getValue(credsInputSchema.discriminator, nodeInputValues)
+      : null,
+    ...(credsInputSchema.discriminator_values || []),
+  ].find(Boolean);
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/hooks/useCredentials.ts:46]: what does this `.find(Boolean)` do?

File: autogpt_platform/backend/backend/blocks/__init__.py [hunk 17]
```diff
@@ -20,7 +20,7 @@ def load_all_blocks() -> dict[str, type["Block"]]:
     modules = [
         str(f.relative_to(current_dir))[:-3].replace(os.path.sep, ".")
         for f in current_dir.rglob("*.py")
-        if f.is_file() and f.name != "__init__.py"
+        if f.is_file() and f.name != "__init__.py" and not f.name.startswith("test_")
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/blocks/__init__.py:23]: maybe also include a pattern for `*_test.py`?
```suggestion
        if (
            f.is_file()
            and f.name != "__init__.py"
            and not f.name.startswith("test_")
            and not f.name.endswith("_test.py")
        )
```

File: autogpt_platform/backend/backend/data/graph.py [hunk 18]
```diff
@@ -243,6 +244,8 @@ def _credentials_input_schema(self) -> type[BlockSchema]:
             for other_field, other_keys in list(graph_cred_fields)[i + 1 :]:
                 if field.provider != other_field.provider:
                     continue
+                if ProviderName.HTTP in field.provider:
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/graph.py:247]:
```suggestion
                if field.provider == ProviderName.HTTP:
```

File: autogpt_platform/backend/backend/data/model.py [hunk 19]
```diff
@@ -320,27 +373,42 @@ def validate_credentials_field_schema(cls, model: type["BlockSchema"]):
...
-    def _add_json_schema_extra(schema, cls: CredentialsMetaInput):
-        schema["credentials_provider"] = cls.allowed_providers()
-        schema["credentials_types"] = cls.allowed_cred_types()
+    def _add_json_schema_extra(schema: dict, model_class: type):
+        # Use model_class for allowed_providers/cred_types
+        if hasattr(model_class, "allowed_providers") and hasattr(
+            model_class, "allowed_cred_types"
+        ):
+            schema["credentials_provider"] = model_class.allowed_providers()
+            schema["credentials_types"] = model_class.allowed_cred_types()
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model.py:382]: why this change?

File: autogpt_platform/backend/backend/data/model.py [hunk 20]
```diff
@@ -358,22 +426,36 @@ def combine(
...
+                )
+            else:
+                providers = frozenset(field.provider)
+
+            group_key = (providers, frozenset(field.supported_types))
             grouped_fields[group_key].append((key, field))
 
         # Combine fields within each group
-        result: list[tuple[CredentialsFieldInfo[CP, CT], set[T]]] = []
+        result: dict[str, tuple[CredentialsFieldInfo[CP, CT], set[T]]] = {}
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model.py:456]: should have updated the docstring as well

File: autogpt_platform/backend/backend/data/model.py [hunk 21]
```diff
@@ -386,18 +468,32 @@ def combine(
...
-                        discriminator_mapping=combined.discriminator_mapping,
-                    ),
-                    combined_keys,
-                )
+            # Combine discriminator_values from all fields in the group (removing duplicates)
+            all_discriminator_values = []
+            for _, field in group:
+                for value in field.discriminator_values:
+                    if value not in all_discriminator_values:
+                        all_discriminator_values.append(value)
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model.py:476]: nit: could use a set for this

File: autogpt_platform/backend/backend/data/model.py [hunk 22]
```diff
@@ -320,27 +373,42 @@ def validate_credentials_field_schema(cls, model: type["BlockSchema"]):
...
+
+
 class CredentialsFieldInfo(BaseModel, Generic[CP, CT]):
     # TODO: move discrimination mechanism out of CredentialsField (frontend + backend)
     provider: frozenset[CP] = Field(..., alias="credentials_provider")
     supported_types: frozenset[CT] = Field(..., alias="credentials_types")
     required_scopes: Optional[frozenset[str]] = Field(None, alias="credentials_scopes")
     discriminator: Optional[str] = None
     discriminator_mapping: Optional[dict[str, CP]] = None
+    discriminator_values: set[Any] = Field(default_factory=set)
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model.py:406]: Can there be multiple discriminator values for a single node? When/how would that happen?

File: autogpt_platform/backend/backend/data/model_test.py [hunk 23]
```diff
@@ -0,0 +1,143 @@
+import pytest
+from pydantic import SecretStr
+
+from backend.data.model import HostScopedCredentials
+
+
+class TestHostScopedCredentials:
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/model_test.py:7]: another file with class-wrapped instead of module-level tests :)

File: autogpt_platform/backend/backend/blocks/test/test_http.py [hunk 24]
```diff
@@ -0,0 +1,485 @@
...
+from backend.blocks.http import (
+    HttpCredentials,
+    HttpMethod,
+    SendAuthenticatedWebRequestBlock,
+)
+from backend.data.model import HostScopedCredentials
+from backend.util.request import Response
+
+
+class TestHttpBlockWithHostScopedCredentials:
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/blocks/test/test_http.py:18]: another one
